### PR TITLE
[4.0] Allow empty metakey values in database tables

### PR DESCRIPTION
**Pull request by richard67**
Pull Request for [https://github.com/joomla/joomla-cms/pull/25258#issuecomment-583731167](https://github.com/joomla/joomla-cms/pull/25258#issuecomment-583731167).

### Summary of Changes

With PR #25258 meta keywords have been removed from diverse forms and views.

This is a good thing.

But the fields in database tables still exist, have a `NOT NULL` constraint and partly no default value, e.g. in the `#__languages` table.

This leads to content language not being installed when installing a language, see [https://github.com/joomla/joomla-cms/pull/25258#issuecomment-583731167](https://github.com/joomla/joomla-cms/pull/25258#issuecomment-583731167).

Furthermore I saw fields `metakey` are still used by diverse sample data, so we can't just remove those columns, not yet.

So this PR removes the `NOT NULL` constraint from `metakey` columns where they are of type `text`, because for such columns default values are not allowed in strict mode, and adds missing default value empty string to those columns where they are of type `VARCHAR`, so things don't fail and stay BC.

Later when all metaky removal has been completed we can remove those columns from database tables with another PR, wherever appropriate (BC?).

#### Update

1. On a current 4.0-dev installation, apply the changes in the update sql script `4.0.0-2020-02-08.sql` added by this PR e.g. with PhpMyAdmin or PhpPgAdmin, replacing `#___` by your actual db prefix.

2. Install Persian language.
**Result:** Content language created, no notice like shown in comment [https://github.com/joomla/joomla-cms/pull/25258#issuecomment-583731167](https://github.com/joomla/joomla-cms/pull/25258#issuecomment-583731167).

3. Verify that following still works:
- Creating and editing banners and banner clients
- Creating and editing contacts
- Creating and editing articles
- Creating and editing newsfeeds
- Creating and editing tags

#### New installation

1. Apply PR on a clean current 4.0-dev and then make new installation.

2. to 3. Same as for update above.

### Expected result

Code review/database check: All database columns for meta keywords have default value `''` where they are of type `VARCHAR` and have no default and allow `NULL` values where they are of type `TEXT`.

Language installation test (step 2): No notice, and content language is created.

Other tests (step 3): Things work.

### Actual result

Code review/database check: There are some database columns for meta keywords having a `NOT NULL` constraint and no default value.

Language installation test (step 2): Notice shown, and content language not created.

j4-install-persian-language-notice

Other tests (step 3): Things work.

### Documentation Changes Required

None.

### CODE REVIEW

File: installation/sql/mysql/joomla.sql [hunk 1]
```diff
@@ -1763,7 +1763,7 @@ CREATE TABLE IF NOT EXISTS `#__tags` (
   `access` int(10) unsigned NOT NULL DEFAULT 0,
   `params` text NOT NULL,
   `metadesc` varchar(1024) NOT NULL COMMENT 'The meta description for the page.',
-  `metakey` varchar(1024) NOT NULL COMMENT 'The meta keywords for the page.',
+  `metakey` varchar(1024) NOT NULL DEFAULT '' COMMENT 'The meta keywords for the page.',
```
Comment (brianteeman) [inline installation/sql/mysql/joomla.sql]: i would probablt drop the word meta from the comment

Comment (richard67) [reply installation/sql/mysql/joomla.sql]: shit for fields of type text default '' is not allowed in mysql

Comment (richard67) [reply installation/sql/mysql/joomla.sql]: it seems this whole pr is shit

Comment (richard67) [reply installation/sql/mysql/joomla.sql]: wait maybe only update sql is wrong .. anyway, have to fix

Comment (richard67) [reply installation/sql/mysql/joomla.sql]: Ah no it was only mysql strict mode, which is switched off anyway when the update sql runs. I just have to update testing instructions for it. then I'll apply your suggested changes.

**Comment by richard67** [comment]
@brianteeman Done. I won't change the comments with update SQL scripts because we never do that for comments only.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 2723a32b1806d081bbf6b3c18bb588f392ecafa3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27854">issues.joomla.org/tracker/joomla-cms/27854</a>.</sub>

**Comment by richard67** [comment]
Thinking about it ... maybe I should change this to null values instead of empty strings? Is late here, I have to check tomorrow.

**Comment by brianteeman** [comment]
I did wonder about that but you're the sql guru

**Comment by alikon** [comment]
from https://dev.mysql.com/doc/refman/8.0/en/blob.html

> BLOB and TEXT columns cannot have DEFAULT values.

**Comment by richard67** [comment]
You shall not trust gurus 😝

**Comment by richard67** [comment]
I should not have looked into joomla.sql to see how it should be done ... is all wrong there. There is a PR which wants to fix that for staging, #17860 , but it it missing something, and the author seems not to be available anymore. Have pinged him yesterday. Let's wait a bit and give him time. For this PR here I will check if I can provide a correction, and if not I'll close it and make an issue.

@Quy Thanks for pointing me to that other PR. Clould you set back your test result to "not tested" for my PR here?

**Comment by richard67** [comment]
@alikon @brianteeman @Quy PR modified, descriptions and testing instructions updated. Ready for testing.

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on f08bcf414fd063ada5d95ac593c406a1694c39bc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27854">issues.joomla.org/tracker/joomla-cms/27854</a>.</sub>

**Comment by infograf768** [comment]
rather urgent to get in as multilingual is dead fish. ;)

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on f08bcf414fd063ada5d95ac593c406a1694c39bc<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27854">issues.joomla.org/tracker/joomla-cms/27854</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27854">issues.joomla.org/tracker/joomla-cms/27854</a>.</sub>

**Pull request merged**